tools/analyzer_plugin/lib/src/diagnostic/create_ref_usage.dart

```diff
@@ -0,0 +1,96 @@
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer_plugin/protocol/protocol_common.dart';
+import 'package:over_react_analyzer_plugin/src/diagnostic_contributor.dart';
+import 'package:over_react_analyzer_plugin/src/util/function_components.dart';
+import 'package:over_react_analyzer_plugin/src/util/react_types.dart';
+
+const _desc = r'Prefer useRef over createRef within a function component.';
+// <editor-fold desc="Documentation Details">
+const _details = r'''
+
+This diagnostic detects when `createRef`, instead of `useRef`, is used inside a function component.
+
+**BAD:** 
+```
+import 'package:over_react/over_react.dart';
+
+part 'create_ref_usage.over_react.g.dart';
+
+UiFactory<FooProps> Foo = uiFunction(
+  (props) {
+    final ref = createRef();
+
+    return (Dom.div()..ref = ref)();
+  },
+  _$FooConfig, // ignore: undefined_identifier
+);
+
+mixin FooProps on UiProps {}
+```
+
+**GOOD:**
+```
+import 'package:over_react/over_react.dart';
+
+part 'create_ref_usage.over_react.g.dart';
+
+UiFactory<FooProps> Foo = uiFunction(
+  (props) {
+    final ref = useRef();
+
+    return (Dom.div()..ref = ref)();
+  },
+  _$FooConfig, // ignore: undefined_identifier
+);
+
+mixin FooProps on UiProps {}
+```
+
+''';
+// </editor-fold>
+
+class CreateRefUsageDiagnostic extends DiagnosticContributor {
+  @DocsMeta(_desc, details: _details)
+  static const code = DiagnosticCode(
+    'over_react_create_ref_usage',
+    "Prefer useRef over createRef within a function component.",
+    AnalysisErrorSeverity.WARNING,
+    AnalysisErrorType.STATIC_WARNING,
+  );
+
+  static final fixKind = FixKind(code.name, 400, "Replace with useRef");
+
+  @override
+  computeErrors(result, collector) async {
+    result.unit!.accept(FunctionComponentVisitor((component) async {
+      final visitor = CreateRefUsageVisitor();
+      component.body.accept(visitor);
+      for (final createRefInvocation in visitor.createRefInvocations.keys) {
+        await collector.addErrorWithFix(
+          code,
+          result.locationFor(createRefInvocation),
+          fixKind: fixKind,
+          computeFix: () => buildGenericFileEdit(result, (builder) {
+            builder.addSimpleReplacement(range.node(createRefInvocation), 'useRef');
+          }),
+        );
+      }
+    }));
+  }
+}
+
+class CreateRefUsageVisitor extends GeneralizingAstVisitor<void> {
+  final createRefInvocations = <SimpleIdentifier, Uri>{};
```

Comment:
Is there a reason to store the `Uri`? I don't see it used anywhere.

Would a `Set` be sufficent?

Reply:
Oh good catch! I was previously doing something with the Uri, but now that's unnecessary - I'll remove that!

Reply:
I had another thought. Maybe we shouldn't be using `SimpleIdentifier` as a key since it doesn't implement a custom `hashCode`.